lib/routes/token-codes.js

```diff
@@ -0,0 +1,66 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const errors = require('../error')
+const isA = require('joi')
+const validators = require('./validators')
+const HEX_STRING = validators.HEX_STRING
+
+module.exports = (log, db, customs) => {
+  return [
+    {
+      method: 'POST',
+      path: '/session/verify/token',
+      config: {
+        auth: {
+          strategy: 'sessionToken'
+        },
+        validate: {
+          payload: {
+            uid: isA.string().max(32).regex(HEX_STRING).required(),
+            code: isA.string().min(0).max(1024).required() // code length is configurable but lets add some sane min/max values
+          }
+        }
+      },
+      handler (request, reply) {
+        log.begin('session.verify.token', request)
+
+        const uid = request.payload.uid
+        const code = request.payload.code.toUpperCase()
+
+        customs.checkIpOnly(request, 'verifyTokenCode')
+          .then(verifyCode)
+          .then(emitMetrics)
+          .then(reply, reply)
+
+        function verifyCode() {
+          return db.verifyTokenCode(code, {uid: uid})
+            .then(() => {}, (err) => {
+              if (err.errno === errors.ERRNO.EXPIRED_TOKEN_VERIFICATION_CODE) {
+                log.error({
+                  op: 'account.token.code.expired',
+                  uid: uid,
+                  err: err
+                })
+              }
+              throw err
+            })
+        }
+
+        function emitMetrics() {
+          log.info({
+            op: 'account.token.code.verified',
+            uid: uid
+          })
+
+          return request.emitMetricsEvent('tokenCodes.verified', {uid: uid})
+            .then(() => ({}))
+        }
```

Comment:
I suspect we will need to emit an `account.confirmed` metrics event here, for at least two reasons:

* This event is used by amplitude to generate `fxa_login - email_confirmed` events, so users in this experiment might otherwise disappear from our login funnels in amplitude
* This is used as the flow-complete signal for non-sync logins, so users in this experiment might incorrectly appear to have failed to complete their login flow.